File: 941. Valid Mountain Array.cpp
class Solution {
public:
    bool validMountainArray(vector<int>& arr) {
        int n = arr.size();
        if (n < 3)
            return false;
        unordered_map<string, int> stateChanges;
        bool isUp = false;
        bool isDown = false;

        for (int i = 1; i < n; i++) {
            if (arr[i] > arr[i - 1]) {
                if (isDown)
                    return false;
                isUp = true;
            } else if (arr[i] < arr[i - 1]) {
                if (!isUp)
                    return false;
                isDown = true;
            } else {
                return false;
            }
        }

        return isUp && isDown;
    }
};

File: 941. Valid Mountain Array.java
class Solution {
    public boolean validMountainArray(int[] arr) {
        int n = arr.length, i = 0, j = n - 1;
        if (n < 3) {
            return false;
        }
        while ((i + 1) < (n - 1) && arr[i] < arr[i + 1]) {
            i++;
        }
        while (j > 0 && arr[j - 1] > arr[j]) {
            j--;
        }
        return i > 0 && i == j && j < n - 1;
    }
}
